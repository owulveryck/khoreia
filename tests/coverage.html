
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">github.com/owulveryck/gchoreography/choreography/executor.go (90.9%)</option>
				
				<option value="file1">github.com/owulveryck/gchoreography/choreography/graph.go (95.0%)</option>
				
				<option value="file2">github.com/owulveryck/gchoreography/choreography/node.go (81.2%)</option>
				
				<option value="file3">github.com/owulveryck/gchoreography/choreography/structure.go (100.0%)</option>
				
				<option value="file4">github.com/owulveryck/gchoreography/choreography/util.go (100.0%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" >package choreography

import (
        "crypto/tls"
        "crypto/x509"
        "fmt"
        "io/ioutil"
        "net/http"
)

// ExecutorBackend represent an executor
type ExecutorBackend struct {
        Url         string // eg: https://localhost:8585/v1
        Certificate string //
        Key         string
        CACert      string
        Ping        string // eg /ping
        Client      *http.Client
}

func (e *ExecutorBackend) Init() error <span class="cov8" title="1">{
        var client *http.Client
        // Load client cert
        cert, err := tls.LoadX509KeyPair(e.Certificate, e.Key)
        if err != nil </span><span class="cov8" title="1">{
                return err
        }</span>

        // Load CA cert
        <span class="cov8" title="1">caCert, err := ioutil.ReadFile(e.CACert)
        if err != nil </span><span class="cov8" title="1">{
                return err

        }</span>
        <span class="cov8" title="1">caCertPool := x509.NewCertPool()
        r := caCertPool.AppendCertsFromPEM(caCert)
        if r == false </span><span class="cov0" title="0">{
                return fmt.Errorf("No certificate found in %v", e.CACert)
        }</span>

        // Setup HTTPS client
        <span class="cov8" title="1">tlsConfig := &amp;tls.Config{
                Certificates: []tls.Certificate{cert},
                RootCAs:      caCertPool,
        }
        tlsConfig.BuildNameToCertificate()
        transport := &amp;http.Transport{TLSClientConfig: tlsConfig}
        client = &amp;http.Client{Transport: transport}

        e.Client = client
        // Now check the ping url
        resp, err := client.Get(fmt.Sprintf("%v%v", e.Url, e.Ping))
        if err != nil </span><span class="cov0" title="0">{
                return err
        }</span>
        <span class="cov8" title="1">if resp.StatusCode &gt; 200 </span><span class="cov8" title="1">{
                return fmt.Errorf("Ping not found")
        }</span>

        <span class="cov8" title="1">return nil</span>
}
</pre>
		
		<pre class="file" id="file1" style="display: none">/*
Olivier Wulveryck - author of Gchoreography
Copyright (C) 2015 Olivier Wulveryck

This file is part of the Gchoreography project and
is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.
*/
package choreography

import (
        "github.com/owulveryck/gchoreography/structure"
        "sync"
        "time"
)

// Graph is the input of the choreography
type Graph struct {
        Name    string           `json:"name,omitempty"`
        State   int              `json:"state"`
        Digraph structure.Matrix `json:"digraph"`
        Nodes   []Node           `json:"nodes"`
        Timeout &lt;-chan time.Time `json:"-"`
}

func (v *Graph) getNodeFromName(n string) (Node, error) <span class="cov0" title="0">{
        var nn Node
        return nn, nil
}</span>

var mu sync.RWMutex

// Run executes the Graph structure
func (v *Graph) Run(exe ExecutorBackend) <span class="cov8" title="1">{
        v.State = Running
        m := v.Digraph

        n := m.Dim()
        cs := make([]&lt;-chan Message, n)
        cos := make([]chan&lt;- Graph, n)
        for i := 0; i &lt; n; i++ </span><span class="cov8" title="1">{
                cs[i] = v.Nodes[i].Run(exe)
        }</span>
        <span class="cov8" title="1">for i := 0; i &lt; n; i++ </span><span class="cov8" title="1">{
                node := &lt;-cs[i]
                cos[i] = node.Wait
        }</span>

        <span class="cov8" title="1">co := fanOut(cos...)
        c := fanIn(cs...)
        for </span><span class="cov8" title="1">{
                select </span>{
                <span class="cov8" title="1">case node := &lt;-c:
                        if node.State &gt;= Running </span><span class="cov8" title="1">{
                                for c := 0; c &lt; n; c++ </span><span class="cov8" title="1">{
                                        if m.At(node.ID, c) != 0 </span><span class="cov8" title="1">{
                                                mu.Lock()
                                                m.Set(node.ID, c, int64(node.State))
                                                mu.Unlock()
                                        }</span>
                                }
                        }
                        <span class="cov8" title="1">stop := true
                        v.State = Success
                        for r := 0; r &lt; n; r++ </span><span class="cov8" title="1">{
                                for c := 0; c &lt; n; c++ </span><span class="cov8" title="1">{
                                        switch </span>{
                                        <span class="cov8" title="1">case m.At(r, c) == ToRun:
                                                stop = false
                                                v.State = Running</span>
                                        <span class="cov8" title="1">case m.At(r, c) == Running:
                                                stop = false
                                                v.State = Running</span>
                                        <span class="cov8" title="1">case m.At(r, c) &gt; Success:
                                                v.State = Failure</span>
                                        }
                                }
                        }
                        <span class="cov8" title="1">if stop </span><span class="cov8" title="1">{
                                return
                        }</span>
                <span class="cov8" title="1">case &lt;-v.Timeout:
                        co &lt;- Graph{
                                (*v).Name,
                                (*v).State,
                                (*v).Digraph,
                                (*v).Nodes,
                                (*v).Timeout,
                        }
                        v.State = Timeout
                        return</span>
                <span class="cov8" title="1">case co &lt;- Graph{
                        (*v).Name,
                        (*v).State,
                        (*v).Digraph,
                        (*v).Nodes,
                        (*v).Timeout,
                }:</span>
                }
        }
}

// Check is the structure is coherent, (a squared matrix with as many nodes as needed)
func (i *Graph) Check() Error <span class="cov8" title="1">{
        if len(i.Nodes)*len(i.Nodes) != len(i.Digraph) </span><span class="cov8" title="1">{
                return Error{1, "Structure is not coherent"}
        }</span>
        <span class="cov8" title="1">return Error{0, ""}</span>
}
</pre>
		
		<pre class="file" id="file2" style="display: none">/*
Olivier Wulveryck - author of Gchoreography
Copyright (C) 2015 Olivier Wulveryck

This file is part of the Gchoreography project and
is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.
*/

package choreography

import (
        "bytes"
        "encoding/json"
        "errors"
        "fmt"
        "github.com/owulveryck/gchoreography/structure"
        "math/rand"
        "net/http"
        "regexp"
        "time"
)

// Node is a "runable" node description
type Node struct {
        ID       int               `json:"id"`
        State    int               `json:"state,omitempty"`
        Name     string            `json:"name,omitempty"`
        Engine   string            `json:"engine,omitempty"` // The execution engine (ie ansible, shell); aim to be like a shebang in a shell file
        Artifact string            `json:"artifact"`
        Args     []string          `json:"args,omitempty"`   // the arguments of the artifact, if needed
        Outputs  map[string]string `json:"output,omitempty"` // the key is the name of the parameter, the value its value (always a string)
}

// Actually executes the node (via the executor)
func (n *Node) Execute(exe ExecutorBackend) error <span class="cov8" title="1">{
        if exe.Client == nil </span><span class="cov0" title="0">{
                err := exe.Init()
                if err != nil </span><span class="cov0" title="0">{
                        return err
                }</span>
        }
        <span class="cov8" title="1">var id string
        var err error
        var t struct {
                ID string `json:"id"`
        }
        url := fmt.Sprintf("%v/tasks", exe.Url)
        b, _ := json.Marshal(*n)
        req, err := http.NewRequest("POST", url, bytes.NewBuffer(b))
        //req.Header.Set("X-Custom-Header", "myvalue")
        req.Header.Set("Content-Type", "application/json")

        //client := &amp;http.Client{}
        client := exe.Client
        // Do a ping before for testing purpose
        resp, err := client.Do(req)
        if err != nil </span><span class="cov0" title="0">{
                n.State = NotRunnable
                return err

        }</span>
        <span class="cov8" title="1">defer resp.Body.Close()

        if resp.StatusCode &gt;= http.StatusBadRequest </span><span class="cov0" title="0">{
                n.State = NotRunnable
                return errors.New("Error in the executor")

        }</span>
        <span class="cov8" title="1">dec := json.NewDecoder(resp.Body)
        if err := dec.Decode(&amp;t); err != nil </span><span class="cov0" title="0">{
                n.State = Failure
                return err
        }</span>
        <span class="cov8" title="1">id = t.ID

        // Now loop for the result
        for err == nil &amp;&amp; n.State &lt; Success </span><span class="cov8" title="1">{
                var res Node
                r, err := client.Get(fmt.Sprintf("%v/%v", url, id))
                if err != nil </span><span class="cov0" title="0">{
                        n.State = NotRunnable
                        return err
                }</span>
                <span class="cov8" title="1">defer r.Body.Close()
                dec := json.NewDecoder(r.Body)
                if err := dec.Decode(&amp;res); err != nil </span><span class="cov0" title="0">{
                        n.State = Failure
                        return err
                }</span>
                <span class="cov8" title="1">*n = res
                time.Sleep(1 * time.Second)</span>
        }
        <span class="cov8" title="1">return nil</span>
}

// Run the node
func (n *Node) Run(exe ExecutorBackend) &lt;-chan Message <span class="cov8" title="1">{
        c := make(chan Message)
        waitForIt := make(chan Graph) // Shared between all messages.
        var ga = regexp.MustCompile(`^get_attribute (.+):(.+)$`)

        var g Graph
        go func() </span><span class="cov8" title="1">{
                n.State = ToRun
                for n.State &lt;= ToRun </span><span class="cov8" title="1">{
                        c &lt;- Message{n.ID, n.State, waitForIt}
                        g = &lt;-waitForIt
                        var m structure.Matrix
                        m = g.Digraph
                        s := m.Dim()
                        n.Outputs = make(map[string]string, 0)
                        if n.Artifact == "" &amp;&amp; n.Engine == "" </span><span class="cov8" title="1">{
                                n.Engine = "nil"
                        }</span>
                        <span class="cov8" title="1">n.State = Running
                        for i := 0; i &lt; s; i++ </span><span class="cov8" title="1">{
                                mu.RLock()
                                if m.At(i, n.ID) &lt; Success &amp;&amp; m.At(i, n.ID) &gt; 0 </span><span class="cov8" title="1">{
                                        n.State = ToRun
                                }</span><span class="cov8" title="1"> else if m.At(i, n.ID) &gt;= Failure </span><span class="cov8" title="1">{
                                        n.State = NotRunnable
                                }</span>
                                <span class="cov8" title="1">mu.RUnlock()
                                if n.State == NotRunnable </span><span class="cov8" title="1">{
                                        continue</span>
                                }
                        }
                        <span class="cov8" title="1">if n.State == NotRunnable </span><span class="cov8" title="1">{
                                c &lt;- Message{n.ID, n.State, waitForIt}
                        }</span>
                        <span class="cov8" title="1">if n.State == Running </span><span class="cov8" title="1">{
                                // Check and find the arguments
                                for i, arg := range n.Args </span><span class="cov8" title="1">{
                                        // If argument is a get_attribute node:attribute
                                        // Then substitute it to its actual value
                                        subargs := ga.FindStringSubmatch(arg)
                                        if len(subargs) == 4 </span><span class="cov0" title="0">{
                                                nn, _ := g.getNodeFromName(subargs[2])
                                                n.Args[i] = nn.Outputs[subargs[3]]
                                        }</span>
                                }
                                <span class="cov8" title="1">c &lt;- Message{n.ID, n.State, waitForIt}
                                switch n.Engine </span>{
                                <span class="cov8" title="1">case "nil":
                                        n.State = Success</span>
                                <span class="cov8" title="1">case "sleep": // For test purpose
                                        time.Sleep(time.Duration(rand.Intn(1e3)) * time.Millisecond)
                                        rand.Seed(time.Now().Unix())
                                        n.State = Success
                                        n.Outputs["result"] = fmt.Sprintf("%v_%v", n.Name, time.Now().Unix())</span>
                                <span class="cov8" title="1">default:
                                        err := n.Execute(exe)
                                        if err != nil &amp;&amp; n.State &lt;= Success </span><span class="cov0" title="0">{
                                                n.State = Failure
                                        }</span>

                                }
                                <span class="cov8" title="1">c &lt;- Message{n.ID, n.State, waitForIt}</span>
                        }
                }
                <span class="cov8" title="1">close(c)</span>
        }()
        <span class="cov8" title="1">return c</span>
}
</pre>
		
		<pre class="file" id="file3" style="display: none">/*
Olivier Wulveryck - author of Gchoreography
Copyright (C) 2015 Olivier Wulveryck

This file is part of the Gchoreography project and
is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.
*/

package choreography

import (
        "encoding/json"
)

type Message struct {
        ID    int
        State int
        Wait  chan Graph
}

// Error is a type used when any error related to the input or node structure occurs
type Error struct {
        Code    int    `json:"code"`
        Message string `json:"message"`
}

func (e *Error) Error() string <span class="cov8" title="1">{
        o, _ := json.Marshal(e)
        return string(o)
}</span>
</pre>
		
		<pre class="file" id="file4" style="display: none">/*
Olivier Wulveryck - author of Gchoreography
Copyright (C) 2015 Olivier Wulveryck

This file is part of the Gchoreography project and
is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.
*/

package choreography

func fanOut(outputs ...chan&lt;- Graph) chan&lt;- Graph <span class="cov8" title="1">{
        c := make(chan Graph)
        for i := range outputs </span><span class="cov8" title="1">{
                output := outputs[i] // New instance of 'input' for each loop.
                go func(i int) </span><span class="cov8" title="1">{
                        for </span><span class="cov8" title="1">{
                                output &lt;- &lt;-c
                        }</span>
                }(i)
        }
        <span class="cov8" title="1">return c</span>
}
func fanIn(inputs ...&lt;-chan Message) &lt;-chan Message <span class="cov8" title="1">{
        c := make(chan Message)
        for i := range inputs </span><span class="cov8" title="1">{
                input := inputs[i] // New instance of 'input' for each loop.
                go func() </span><span class="cov8" title="1">{
                        for </span><span class="cov8" title="1">{
                                c &lt;- &lt;-input
                        }</span>
                }()
        }
        <span class="cov8" title="1">return c</span>
}
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible = document.getElementById('file0');
		files.addEventListener('change', onChange, false);
		function onChange() {
			visible.style.display = 'none';
			visible = document.getElementById(files.value);
			visible.style.display = 'block';
			window.scrollTo(0, 0);
		}
	})();
	</script>
</html>
