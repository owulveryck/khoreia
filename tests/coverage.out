mode: set
github.com/owulveryck/gchoreography/choreography/executor.go:21.40,25.16 3 1
github.com/owulveryck/gchoreography/choreography/executor.go:30.2,31.16 2 1
github.com/owulveryck/gchoreography/choreography/executor.go:35.2,37.16 3 1
github.com/owulveryck/gchoreography/choreography/executor.go:42.2,53.16 7 1
github.com/owulveryck/gchoreography/choreography/executor.go:56.2,56.27 1 1
github.com/owulveryck/gchoreography/choreography/executor.go:60.2,60.12 1 1
github.com/owulveryck/gchoreography/choreography/executor.go:25.16,27.3 1 1
github.com/owulveryck/gchoreography/choreography/executor.go:31.16,34.3 1 1
github.com/owulveryck/gchoreography/choreography/executor.go:37.16,39.3 1 0
github.com/owulveryck/gchoreography/choreography/executor.go:53.16,55.3 1 0
github.com/owulveryck/gchoreography/choreography/executor.go:56.27,58.3 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:36.57,39.2 2 0
github.com/owulveryck/gchoreography/choreography/graph.go:44.42,51.25 6 1
github.com/owulveryck/gchoreography/choreography/graph.go:54.2,54.25 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:59.2,61.6 3 1
github.com/owulveryck/gchoreography/choreography/graph.go:51.25,53.3 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:54.25,57.3 2 1
github.com/owulveryck/gchoreography/choreography/graph.go:61.6,62.10 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:63.3,64.29 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:73.4,75.27 3 1
github.com/owulveryck/gchoreography/choreography/graph.go:89.4,89.12 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:92.3,101.10 3 1
github.com/owulveryck/gchoreography/choreography/graph.go:102.3,108.5 0 1
github.com/owulveryck/gchoreography/choreography/graph.go:64.29,65.28 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:65.28,66.31 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:66.31,70.7 3 1
github.com/owulveryck/gchoreography/choreography/graph.go:75.27,76.28 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:76.28,77.13 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:78.6,80.24 2 1
github.com/owulveryck/gchoreography/choreography/graph.go:81.6,83.24 2 1
github.com/owulveryck/gchoreography/choreography/graph.go:84.6,85.24 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:89.12,91.5 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:114.31,115.49 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:118.2,118.21 1 1
github.com/owulveryck/gchoreography/choreography/graph.go:115.49,117.3 1 1
github.com/owulveryck/gchoreography/choreography/node.go:46.51,47.23 1 1
github.com/owulveryck/gchoreography/choreography/node.go:53.2,68.16 10 1
github.com/owulveryck/gchoreography/choreography/node.go:73.2,75.46 2 1
github.com/owulveryck/gchoreography/choreography/node.go:80.2,81.39 2 1
github.com/owulveryck/gchoreography/choreography/node.go:85.2,88.38 2 1
github.com/owulveryck/gchoreography/choreography/node.go:104.2,104.12 1 1
github.com/owulveryck/gchoreography/choreography/node.go:47.23,49.17 2 0
github.com/owulveryck/gchoreography/choreography/node.go:49.17,51.4 1 0
github.com/owulveryck/gchoreography/choreography/node.go:68.16,72.3 2 0
github.com/owulveryck/gchoreography/choreography/node.go:75.46,79.3 2 0
github.com/owulveryck/gchoreography/choreography/node.go:81.39,84.3 2 0
github.com/owulveryck/gchoreography/choreography/node.go:88.38,91.17 3 1
github.com/owulveryck/gchoreography/choreography/node.go:95.3,97.42 3 1
github.com/owulveryck/gchoreography/choreography/node.go:101.3,102.30 2 1
github.com/owulveryck/gchoreography/choreography/node.go:91.17,94.4 2 0
github.com/owulveryck/gchoreography/choreography/node.go:97.42,100.4 2 0
github.com/owulveryck/gchoreography/choreography/node.go:108.56,114.12 5 1
github.com/owulveryck/gchoreography/choreography/node.go:174.2,174.10 1 1
github.com/owulveryck/gchoreography/choreography/node.go:114.12,116.24 2 1
github.com/owulveryck/gchoreography/choreography/node.go:172.3,172.11 1 1
github.com/owulveryck/gchoreography/choreography/node.go:116.24,123.42 7 1
github.com/owulveryck/gchoreography/choreography/node.go:126.4,127.27 2 1
github.com/owulveryck/gchoreography/choreography/node.go:139.4,139.30 1 1
github.com/owulveryck/gchoreography/choreography/node.go:142.4,142.26 1 1
github.com/owulveryck/gchoreography/choreography/node.go:123.42,125.5 1 1
github.com/owulveryck/gchoreography/choreography/node.go:127.27,129.53 2 1
github.com/owulveryck/gchoreography/choreography/node.go:134.5,135.31 2 1
github.com/owulveryck/gchoreography/choreography/node.go:129.53,131.6 1 1
github.com/owulveryck/gchoreography/choreography/node.go:131.6,131.40 1 1
github.com/owulveryck/gchoreography/choreography/node.go:131.40,133.6 1 1
github.com/owulveryck/gchoreography/choreography/node.go:135.31,136.14 1 1
github.com/owulveryck/gchoreography/choreography/node.go:139.30,141.5 1 1
github.com/owulveryck/gchoreography/choreography/node.go:142.26,144.32 1 1
github.com/owulveryck/gchoreography/choreography/node.go:153.5,154.21 2 1
github.com/owulveryck/gchoreography/choreography/node.go:169.5,169.43 1 1
github.com/owulveryck/gchoreography/choreography/node.go:144.32,148.27 2 1
github.com/owulveryck/gchoreography/choreography/node.go:148.27,151.7 2 0
github.com/owulveryck/gchoreography/choreography/node.go:155.5,156.23 1 1
github.com/owulveryck/gchoreography/choreography/node.go:157.5,161.75 4 1
github.com/owulveryck/gchoreography/choreography/node.go:162.5,164.42 2 1
github.com/owulveryck/gchoreography/choreography/node.go:164.42,166.7 1 0
github.com/owulveryck/gchoreography/choreography/structure.go:38.32,41.2 2 1
github.com/owulveryck/gchoreography/choreography/util.go:22.51,24.25 2 1
github.com/owulveryck/gchoreography/choreography/util.go:32.2,32.10 1 1
github.com/owulveryck/gchoreography/choreography/util.go:24.25,26.18 2 1
github.com/owulveryck/gchoreography/choreography/util.go:26.18,27.8 1 1
github.com/owulveryck/gchoreography/choreography/util.go:27.8,29.5 1 1
github.com/owulveryck/gchoreography/choreography/util.go:34.53,36.24 2 1
github.com/owulveryck/gchoreography/choreography/util.go:44.2,44.10 1 1
github.com/owulveryck/gchoreography/choreography/util.go:36.24,38.13 2 1
github.com/owulveryck/gchoreography/choreography/util.go:38.13,39.8 1 1
github.com/owulveryck/gchoreography/choreography/util.go:39.8,41.5 1 1
